CB-31 Pause Support for Various Conditions

jmittert opened the pull request:
Added functions for checking for Invisibility. They needed an new ExpectedCondition in order to catch ElementNotFoundExceptions and the like.

Eriten commented:
The PR looks correct to me. I assume that you have done some testing; we should write some unit tests to test the invisibility conditions in [`WebDriverElementSystemTest`](https://github.com/sugarcrm/candybean/blob/dev/src/test/java/com/sugarcrm/candybean/automation/webdriver/WebDriverElementSystemTest.java).

It also might be a good idea to implement waitForInvisible in VoodooGrimoire.

jmittert commented:
I added a unit test to WebDriverElementSystemTest. I was just going to copy the pauseUntilVisibleTest, but it looks like the website that it depends on now has an anti bot filter that makes the test fail. Do you think it would be a reasonable idea to write a local webpage with the needed testing elements and test from that instead? If not, I can just fix that test in another commit.

I've also implemented waitForInvisible in Voodoo, but there's no use PRing it until Voodoo has access to the waitForInvisible in Candybean.

jmittert commented:
To complete CB-31, additionally added waitForElement/waitForElementRemoved, waitForAttribute/waitForAttributeRegex, and waitForOnScreen.

dsafar commented on src/main/java/com/sugarcrm/candybean/automation/webdriver/WaitConditions.java:
```diff
@@ -1,10 +1,16 @@
 package com.sugarcrm.candybean.automation.webdriver;
 
+import com.gargoylesoftware.htmlunit.ElementNotFoundException;
```
I don't see this getting used anywhere.  Is it actually needed?

jmittert replied:
I don't remember adding that, so no.

dsafar commented on src/main/java/com/sugarcrm/candybean/automation/webdriver/WaitConditions.java:
```diff
@@ -36,17 +43,16 @@ private static WebDriverElement createWebDriverElement(Hook hook, WebElement ele
-		}
+        List<WebElement> elements = driver.findElements(getBy(hook));
+        if (elements.isEmpty())  throw new CandybeanException("No such elements found");
+        return elements.get(0);
```
The three lines above should be formatted with tabs, not spaces.

dsafar commented on src/main/java/com/sugarcrm/candybean/automation/webdriver/WaitConditions.java:
```diff
@@ -112,10 +116,47 @@ public String toString() {
 	 * Wait until the element is not present on the DOM OR invisible
```
Please JavaDoc why not(visible()) isn't sufficient for this.

dsafar commented on src/main/java/com/sugarcrm/candybean/automation/webdriver/WaitConditions.java:
```diff
@@ -112,10 +116,47 @@ public String toString() {
 	 */
-	public static ExpectedCondition<Boolean> invisible(Hook hook) throws CandybeanException {
-		return ExpectedConditions.invisibilityOfElementLocated(getBy(hook));
+	public static ExpectedCondition<Boolean> invisible(final Hook hook) {
```
Please JavaDoc why not(visible()) isn't sufficient for this.

dsafar commented on src/main/java/com/sugarcrm/candybean/automation/webdriver/WaitConditions.java:
```diff
@@ -337,4 +410,81 @@ public String toString() {
+	 * @return The element if the specified value contains the specified attributed, null otherwise
+	 */
+	public static ExpectedCondition<WebDriverElement> hasAttribute(final Hook hook, final String attribute,
+																   final String value, final boolean expectValue) {
```
Formatting?  Way too much indent on this line.

dsafar commented on src/main/java/com/sugarcrm/candybean/automation/webdriver/WaitConditions.java:
```diff
@@ -337,4 +410,81 @@ public String toString() {
+					should match that, but "starred" should not.
+					 */
+					String[] values = element.getAttribute(attribute).split("\\s");
+					for (int x = 0; x < values.length; x++) {
```
We generally prefer foreach syntax, e.g. "for(String value : values)" for readability.

dsafar commented on src/main/java/com/sugarcrm/candybean/automation/webdriver/WaitConditions.java:
```diff
@@ -337,4 +410,81 @@ public String toString() {
+	 * @throws CandybeanException If the element is not found
+	 */
+	public static ExpectedCondition<WebDriverElement> hasRegexAttribute(final Hook hook, final String attribute,
+																		final String regex, final boolean expectValue)
```
Too much indent again.

jmittert commented:
Fixed whitespace and documentation issues.

dsafar commented on src/test/java/com/sugarcrm/candybean/automation/webdriver/WebDriverElementSystemTest.java:
```diff
@@ -343,6 +374,58 @@ public void getCssValueTest() throws CandybeanException {
+		Assert.assertTrue(iface.getPause().waitForAttribute(new Hook(Strategy.ID, "p1"), "class", "hidden", true, 10) != null);
+		Assert.assertTrue(iface.getPause().waitForAttribute(new Hook(Strategy.ID, "p1"), "class", "hidden", false, 10) != null);
+		Assert.assertTrue(iface.getPause().waitForAttribute(new Hook(Strategy.ID, "p1"), "class", "normal", false, 10) != null);
+		Assert.assertTrue(iface.getPause().waitForAttribute(new Hook(Strategy.ID, "p1"), "class", "normal", false, 10) != null);
```
Two lines in a row look identical -- is this intentional, or should one be passing true?

dsafar commented:
Ready for updates and needs a rebase so it can be merged.

jmittert commented:
Updated and rebased.
